### PR TITLE
[Addressing] Removed Address relations and introduced CodeAwareInterface

### PR DESCRIPTION
**Pull request by NeverResponse**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Fixed tickets | #1536, #2561, #3066
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: features/backend/zones.feature [hunk 1]
```diff
@@ -39,6 +39,7 @@ Feature: Zones
 
     Scenario: Creating new zone requires adding at least 1 member
         Given I am on the zone creation page
+          And I fill in "Code" with "EU-Zone"
```
Comment (pjedrzejewski) [inline features/backend/zones.feature]: I'd use EU code here and "European Union" as name, makes more sense. :)

File: src/Sylius/Bundle/AddressingBundle/Behat/AddressingContext.php [hunk 2]
```diff
@@ -119,6 +120,7 @@ public function thereIsZone($name, $type = ZoneInterface::TYPE_COUNTRY, array $m
 
             call_user_func(array($member, 'set'.ucfirst($type)), $zoneable);
 
+            $member->setCode($this->faker->ipv4);
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Behat/AddressingContext.php]: Why?

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryType.php [hunk 3]
```diff
@@ -27,9 +28,7 @@ class CountryType extends AbstractResourceType
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
         $builder
-            ->add('isoName', 'country', array(
-                'label' => 'sylius.form.country.iso_name',
-            ))
+            ->addEventSubscriber(new AddCodeFormSubscriber())
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryType.php]: We could pass here ``country`` to constructor, right?

File: src/Sylius/Bundle/AddressingBundle/Twig/CountryNameExtension.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /**
+     * @param mixed  $country
+     * @param string $locale
+     *
+     * @return string
+     */
+    public function translateCountryIsoCode($country, $locale = null)
+    {   if ($country instanceof CountryInterface) {
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Twig/CountryNameExtension.php]: CS

File: src/Sylius/Bundle/AddressingBundle/Twig/CountryNameExtension.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ */
+class CountryNameExtension extends \Twig_Extension
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getFilters()
+    {
+        return array(
+            new \Twig_SimpleFilter('countryName', array($this, 'translateCountryIsoCode')),
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Twig/CountryNameExtension.php]: Filter should be: ``sylius_country_name``.

File: src/Sylius/Bundle/AddressingBundle/Controller/ZoneMemberController.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ * @author Jan Góralski <jan.goralski@lakion.com>
+ */
+class ZoneMemberController extends ResourceController
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function createNew()
+    {
+        $request = $this->getConfiguration()->getRequest();
```
Comment (Arminek) [inline src/Sylius/Bundle/AddressingBundle/Controller/ZoneMemberController.php]: Consider using ``$request`` as a method parameter ``public function createNew(Request $request)``.

File: features/backend/countries.feature [hunk 7]
```diff
@@ -68,6 +68,15 @@ Feature: Countries and provinces
           And "Łódź" should appear on the page
           And "Lubusz" should appear on the page
 
+    @javascript
+    Scenario: Editing Country with duplicated province
```
Comment (pamil) [inline features/backend/countries.feature]: "Editing country"

File: features/backend/zones.feature [hunk 8]
```diff
@@ -25,47 +25,38 @@ Feature: Zones
...
          Then I should see "There are no zones configured"
 
+    @javascript
     Scenario: Accessing the zone creation form
-        Given I am on the dashboard page
-         When I follow "Zones"
-          And I follow "Create zone"
-         Then I should be on the zone creation page
+        Given I am on the zone index page
+        And I select "Country" from "form_type"
```
Comment (pamil) [inline features/backend/zones.feature]: `form_type` smells bad

File: src/Sylius/Bundle/AddressingBundle/Behat/AddressingContext.php [hunk 9]
```diff
@@ -51,12 +52,12 @@ public function thereIsDisabledCountry($name)
...
     public function thereIsCountry($name, $enabled = true, $provinces = null, $flush = true)
     {
-        $isoName = $this->getCountryCodeByEnglishCountryName($name);
+        $countryCode = $this->getCountryCodeByEnglishCountryName($name);
 
         /* @var $country CountryInterface */
-        if (null === $country = $this->getRepository('country')->findOneBy(array('isoName' => $isoName))) {
+        if (null === $country = $this->getRepository('country')->findOneBy(array('code' => $countryCode))) {
             $country = $this->getFactory('country')->createNew();
-            $country->setIsoName(trim($isoName));
```
Comment (pamil) [inline src/Sylius/Bundle/AddressingBundle/Behat/AddressingContext.php]: :+1:

File: src/Sylius/Bundle/AddressingBundle/Controller/ZoneMemberController.php (hunk 6 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Controller/ZoneMemberController.php]: This is not how it works in ResourceController right now. This is not an action method. :)

File: features/backend/countries.feature [hunk 10]
```diff
@@ -97,11 +106,11 @@ Feature: Countries and provinces
...
-        Given I am on the page of country "France"
-         When I press "delete" near "Toulouse"
-          And I click "delete" from the confirmation modal
-         Then I should still be on the page of country "France"
-          And "Toulouse" should not appear on the page
+        Given I am editing country "France"
+          And I remove the first province
+          And I press "Save changes"
+         Then I should be on the page of country "France"
+          And "Nancy" should not appear on the page
```
Comment (pjedrzejewski) [inline features/backend/countries.feature]: :+1:

File: src/Sylius/Bundle/AddressingBundle/Controller/ZoneController.php [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            $repository,
+            'createPaginator',
+            array($criteria, $sorting)
+        );
+
+        $zones->setCurrentPage($request->get('page', 1), true, true);
+        $zones->setMaxPerPage($this->config->getPaginationMaxPerPage());
+        $pluralName = $this->config->getPluralResourceName();
+
+        $form = $this->createFormBuilder()
```
Comment (Arminek) [inline src/Sylius/Bundle/AddressingBundle/Controller/ZoneController.php]: Why do you need form in index action. Is this something related with filtering?

**Comment by pamil** [comment]
Fixes #3066.

File: src/Sylius/Bundle/AddressingBundle/Controller/ZoneController.php (hunk 11 above)
Comment (NeverResponse) [inline src/Sylius/Bundle/AddressingBundle/Controller/ZoneController.php]: For choosing the type of zone which you want to create.

File: src/Sylius/Bundle/AddressingBundle/Controller/ZoneMemberController.php [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        if (null === $zoneId = $request->get('zoneId')) {
+            throw new NotFoundHttpException('No zone given.');
+        }
+
+        $zone = $this
+            ->getZoneController()
+            ->findOr404($request, array('id' => $zoneId))
+        ;
+
+        $zoneMember = parent::createNew();
```
Comment (pamil) [inline src/Sylius/Bundle/AddressingBundle/Controller/ZoneMemberController.php]: `$this->createNew()`?

File: features/backend/zones.feature (hunk 8 above)
Comment (pjedrzejewski) [inline features/backend/zones.feature]: Yes, I am fine with adding a custom step for that:

``And I want to create a :type zone type`` or something like that.

Comment (NeverResponse) [inline features/backend/zones.feature]: :+1:

File: src/Sylius/Bundle/AddressingBundle/Controller/ZoneController.php [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+            ->setTemplateVar($pluralName)
+            ->setData(array(
+                $pluralName => $zones,
+                'form'      => $form->createView(),
+            ))
+        ;
+
+        return $this->handleView($view);
+    }
+}
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Controller/ZoneController.php]: Definitely would like to avoid overriding ``indexAction``. Can we for now hardcode zone types in the template and just put the select in plain html? I have an idea how to handle common form generation but let's leave that for later. Let me know what do you think. :)

File: src/Sylius/Bundle/AddressingBundle/Controller/ZoneMemberController.php [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    /**
+     * @return ResourceController
+     */
+    protected function getZoneController()
+    {
+        return $this->get('sylius.controller.zone');
+    }
+}
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Controller/ZoneMemberController.php]: Why do we need this controller at all? ZoneMembers will be created only in the Zone form, right?

File: src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactory.php [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace Sylius\Bundle\AddressingBundle\Factory;
+
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Sylius\Component\Resource\Factory\Factory;
+
+/**
+ * @author Jan Góralski <jan.goralski@lakion.com>
+ */
+class ZoneFactory extends Factory
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactory.php]: We should have an interface ``ZoneFactoryInterface``.

File: src/Sylius/Bundle/AddressingBundle/Form/EventListener/BuildAddressFormSubscriber.php [hunk 16]
```diff
@@ -30,22 +31,18 @@ class BuildAddressFormSubscriber implements EventSubscriberInterface
     private $countryRepository;
 
     /**
-     * Form factory.
-     *
      * @var FormFactoryInterface
      */
-    private $factory;
+    private $formFactory;
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/EventListener/BuildAddressFormSubscriber.php]: :+1:

File: src/Sylius/Bundle/AddressingBundle/Form/EventListener/BuildAddressFormSubscriber.php [hunk 17]
```diff
@@ -96,13 +99,17 @@ public function preSubmit(FormEvent $event)
...
+        }
+
+        $country = $this->countryRepository->findOneBy(array('code' => $data['country']));
         if (null === $country) {
             return;
         }
 
         if ($country->hasProvinces()) {
-            $event->getForm()->add($this->factory->createNamed('province', 'sylius_province_choice', null, array(
+            $event->getForm()->add($this->formFactory->createNamed('province', 'sylius_province_code_choice', null, array(
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/EventListener/BuildAddressFormSubscriber.php]: I think this is duplicated with preSetData, can we extract it to private method?

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryCodeChoiceType.php [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        parent::configureOptions($resolver);
+
+        $choices = function (Options $options) {
+            if (null === $options['enabled']) {
+                $countries = $this->repository->findAll();
+            } else {
+                $countries = $this->repository->findBy(array('enabled' => $options['enabled']));
+            }
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryCodeChoiceType.php]: Missing blank line before return statement.

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryCodeChoiceType.php [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        return 'sylius_country_code_choice';
+    }
+
+    /**
+     * @param CountryInterface[] $countries
+     *
+     * @return array
+     */
+    protected function getCountryCodes(array $countries)
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryCodeChoiceType.php]: Let's make it private so that nobody is tempted to extend it. ;P

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceChoiceType.php [hunk 20]
```diff
@@ -40,17 +41,20 @@ public function __construct(RepositoryInterface $repository)
...
-            }
+        $choices = function (Options $options) {
+                if (null === $options['country']) {
+                    $choices = $this->repository->findAll();
+                } else {
+                    $choices = $options['country']->getProvinces();
+                }
 
-            return new ObjectChoiceList($options['country']->getProvinces(), null, array(), null, 'id');
+//          return new ObjectChoiceList($choices, null, array(), null, 'id');
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceChoiceType.php]: To remove.

File: src/Sylius/Bundle/AddressingBundle/spec/Factory/ZoneFactorySpec.php [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+class ZoneFactorySpec extends ObjectBehavior
+{
+    function let()
+    {
+        $this->beConstructedWith(Zone::class);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType(ZoneFactory::class);
```
Comment (Arminek) [inline src/Sylius/Bundle/AddressingBundle/spec/Factory/ZoneFactorySpec.php]: In this case we should use string instead of ``::class``. There was conversation about this here https://github.com/Sylius/Sylius/pull/3736#issuecomment-165732890

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneMemberType.php [hunk 22]
```diff
@@ -12,25 +12,49 @@
...
  */
-abstract class ZoneMemberType extends AbstractResourceType
+class ZoneMemberType extends AbstractResourceType
 {
     /**
      * {@inheritdoc}
      */
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $type = $builder->getOption('zone_type');
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneMemberType.php]: You should have it in ``$options['zone_type']``, no?

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneType.php [hunk 23]
```diff
@@ -74,6 +82,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
     /**
      * {@inheritdoc}
      */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        parent::configureOptions($resolver);
+
+        $resolver->setDefault('options', array('zone_type' => null));
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneType.php]: I think that should be just ``array('zone_type' => null`)``. Actually, any problem with setting country as default zone type? I'd be fine with that. In future where we may consider decoupling Zones that would be wrong but for now that would simplify some stuff.

File: src/Sylius/Bundle/AddressingBundle/spec/Factory/ZoneFactorySpec.php [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    function it_extends_factory()
+    {
+        $this->shouldHaveType(Factory::class);
+    }
+
+    function it_creates_zone_with_type()
+    {
+        $this->createTyped('country')->shouldHaveType(Zone::class);
```
Comment (Arminek) [inline src/Sylius/Bundle/AddressingBundle/spec/Factory/ZoneFactorySpec.php]: Maybe you should consider creating custom matcher something like ``beSameAs``.

File: src/Sylius/Bundle/AddressingBundle/Validator/Constraints/ProvinceAddressConstraintValidator.php [hunk 25]
```diff
@@ -12,17 +12,26 @@
...
  * @author Julien Janvier <j.janvier@gmail.com>
  */
 class ProvinceAddressConstraintValidator extends ConstraintValidator
 {
     /**
+     * @var RepositoryInterface
+     */
+    private $countryRepository;
+
+    public function __construct(RepositoryInterface $countryRepository)
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Validator/Constraints/ProvinceAddressConstraintValidator.php]: Missing docblock.

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml [hunk 26]
```diff
@@ -533,6 +533,8 @@ sylius_rbac:
         sylius.zone.update: Edit zone
         sylius.zone.delete: Delete zone
 
+        sylius.zone_member.delete: Delete zone member
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml]: Let's remove it.

File: src/Sylius/Bundle/WebBundle/Resources/config/routing/backend/zone.yml [hunk 27]
```diff
@@ -11,13 +11,18 @@ sylius_backend_zone_index:
...
     defaults:
         _controller: sylius.controller.zone:createAction
         _sylius:
             template: SyliusWebBundle:Backend/Zone:create.html.twig
             redirect: sylius_backend_zone_show
+            factory:
+                method: createTyped
+                arguments: [$type]
+    requirements:
+        type: country|province|zone
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/WebBundle/Resources/config/routing/backend/zone.yml]: Good idea. :+1:

File: src/Sylius/Bundle/AddressingBundle/spec/Factory/ZoneFactorySpec.php (hunk 21 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/spec/Factory/ZoneFactorySpec.php]: Good catch!

**Comment by michalmarcinkowski** [comment]
@NeverResponse is it ready for review?

**Comment by NeverResponse** [comment]
@michalmarcinkowski sure, go ahead.

File: features/backend/countries.feature [hunk 28]
```diff
@@ -84,6 +93,10 @@ Feature: Countries and provinces
          When I click "edit" near "China"
          Then I should be editing country "China"
 
+    Scenario: Cannot edit country code
+        Given I am editing country "China"
```
Comment (pjedrzejewski) [inline features/backend/countries.feature]: When

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceChoiceType.php [hunk 29]
```diff
@@ -40,17 +41,19 @@ public function __construct(RepositoryInterface $repository)
      */
     public function configureOptions(OptionsResolver $resolver)
     {
-        $choiceList = function (Options $options) {
-            if (null === $options['country']) {
-                return new ObjectChoiceList($this->repository->findAll(), null, array(), null, 'id');
-            }
+        $choices = function (Options $options) {
+                if (null === $options['country']) {
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceChoiceType.php]: One tab too far I think?

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneMemberType.php [hunk 30]
```diff
@@ -12,25 +12,48 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver
+            ->setDefaults(array(
+                'empty_value' => 'sylius.form.zone_member.select',
+                'data_class'  => ZoneMember::class,
+                'zone_type'   => 'country',
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneMemberType.php]: We should use a constant here, wdyt?

File: src/Sylius/Bundle/AddressingBundle/Resources/config/routing/zone.yml [hunk 31]
```diff
@@ -7,9 +7,10 @@ sylius_zone_list:
         _controller: sylius.controller.zone:indexAction
 
 sylius_zone_create:
-    path: /create
+    path: /type_{type}/create
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Resources/config/routing/zone.yml]: Maybe just ``/{type}/create``, wdyt?

File: src/Sylius/Component/Core/Promotion/Checker/ShippingCountryRuleChecker.php [hunk 32]
```diff
@@ -36,7 +51,15 @@ public function isEligible(PromotionSubjectInterface $subject, array $configurat
...
         }
 
-        return $address->getCountry()->getId() === $configuration['country'];
+        $country = $this->countryRepository->findOneBy(array('code' => $address->getCountry()));
+
+        if (!$country instanceof CountryInterface) {
+            return false;
+        }
+
+        $countryId = $country->getId();
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Core/Promotion/Checker/ShippingCountryRuleChecker.php]: This variable is used only once, we can remove it.

File: src/Sylius/Bundle/AddressingBundle/DependencyInjection/Compiler/RegisterZoneFactoryPass.php [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->hasDefinition('sylius.factory.zone')) {
+            return;
+        }
+
+        $baseZoneFactory = $container->getDefinition('sylius.factory.zone');
+        $zoneFactoryDefinition = new Definition(ZoneFactory::class);
```
Comment (Arminek) [inline src/Sylius/Bundle/AddressingBundle/DependencyInjection/Compiler/RegisterZoneFactoryPass.php]: IMO you will prevent registering custom factory, cuz you got hard coded class name here ``ZoneFactory::class``.

Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/DependencyInjection/Compiler/RegisterZoneFactoryPass.php]: You should use ``$baseZoneFactory->getClass();``. It will have the custom factory class in case it has been configured.

**Comment by peteward** [comment]
What's the idea behind changing id-relations from address to province and country with codes?

**Comment by pjedrzejewski** [comment]
@peteward It is much simpler to handle and better performance wise. Currently in Sylius we rarely need any information stored on Country entity so there is no point in performing joins all the time. With ZoneMember simplification this is also much simpler and zone matching should be faster and easier to extend with new levels etc. Does it break anything for you? We want to merge it soon.

**Comment by peteward** [comment]
Yeah it's gonna break tons of stuff but sounds like a good change :+1: , we'll work it out ;)

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php [hunk 34]
```diff
@@ -25,14 +25,14 @@ class CountryChoiceType extends AbstractType
     /**
      * @var RepositoryInterface
      */
-    protected $countryRepository;
+    protected $repository;
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php]: Why renaming this?

Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php]: It is a country choice type and there is no other repository so I guess we can leave it like that, wdyt?

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceChoiceType.php [hunk 35]
```diff
@@ -40,17 +40,19 @@ public function __construct(RepositoryInterface $repository)
      */
     public function configureOptions(OptionsResolver $resolver)
     {
-        $choiceList = function (Options $options) {
+        $choices = function (Options $options) {
             if (null === $options['country']) {
-                return new ObjectChoiceList($this->repository->findAll(), null, array(), null, 'id');
+                $choices = $this->repository->findAll();
+            } else {
+                $choices = $options['country']->getProvinces();
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceChoiceType.php]: I would add a check if `$options['country']` is instanceof `CountryInterface`.

Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceChoiceType.php]: Actually, this should be done via ``$resolver->setAllowedTypes(array('country' => CountryInterface::class));``.

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceCodeChoiceType.php [hunk 36]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        parent::configureOptions($resolver);
+
+        $choices = function (Options $options) {
+            if (null === $options['country']) {
+                $provinces = $this->repository->findAll();
+            } else {
+                $provinces = $options['country']->getProvinces();
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceCodeChoiceType.php]: Same as above.

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneCodeChoiceType.php [hunk 37]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * @author Jan Góralski <jan.goralski@lakion.com>
+ */
+class ZoneCodeChoiceType extends AbstractType
+{
+    /**
+     * @var RepositoryInterface
+     */
+    protected $repository;
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneCodeChoiceType.php]: `$zoneRepository`

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneType.php [hunk 38]
```diff
@@ -48,14 +46,26 @@ public function __construct($dataClass, array $validationGroups, array $scopeCho
...
+                'disabled' => true,
+            ))
+            ->add('members', 'collection', array(
+                'type'             => 'sylius_zone_member',
                 'button_add_label' => 'sylius.zone.add_member',
+                'allow_add'        => true,
+                'allow_delete'     => true,
+                'by_reference'     => false,
+                'delete_empty'     => true,
+                'options'          => array(
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneType.php]: We shouldn't align `=>` vertically.

File: src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/Country.orm.xml [hunk 39]
```diff
@@ -15,14 +15,14 @@
...
-            <index columns="iso_name" />
+            <index columns="code" />
         </indexes>
 
         <id name="id" column="id" type="integer">
             <generator strategy="AUTO" />
         </id>
 
-        <field name="isoName" column="iso_name" type="string" length="3" unique="true"/>
+        <field name="code" column="code" type="string" unique="true" />
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/Country.orm.xml]: `length="3"` should stay and we should also add a constraint for that.

File: src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/Province.orm.xml [hunk 40]
```diff
@@ -14,12 +14,18 @@
...
+            <unique-constraint columns="country_id,name" />
+            <unique-constraint columns="country_id,code" />
+        </unique-constraints>
+
         <id name="id" column="id" type="integer">
             <generator strategy="AUTO" />
         </id>
 
+        <field name="code" column="code" type="string" />
+
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/Province.orm.xml]: Redundant blank line

File: src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/Zone.orm.xml [hunk 41]
```diff
@@ -18,6 +18,8 @@
             <generator strategy="AUTO" />
           </id>
 
+        <field name="code" column="code" type="string" unique="true" />
+
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/Zone.orm.xml]: Redundant blank line

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceChoiceType.php (hunk 35 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceChoiceType.php]: :+1:

File: src/Sylius/Bundle/AddressingBundle/spec/Form/Type/AddressTypeSpec.php [hunk 42]
```diff
@@ -45,48 +45,48 @@ function it_has_a_valid_name()
...
         $builder
             ->add('city', 'text', Argument::any())
-            ->shouldBeCalled()
-            ->willReturn($builder);
+            ->willReturn($builder)
+        ;
 
         $builder
             ->add('postcode', 'text', Argument::any())
-            ->shouldBeCalled()
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/spec/Form/Type/AddressTypeSpec.php]: `shouldBeCalled` should stay when using builders. It's the only way we can check it was really invoked.

File: src/Sylius/Bundle/AddressingBundle/spec/Form/Type/ZoneMemberTypeSpec.php [hunk 43]
```diff
@@ -45,34 +47,29 @@ function it_is_a_Sylius_zone_member_type()
...
     {
-        $this->getName()->shouldReturn('sylius_zone_member_zone');
+        $this->getName()->shouldReturn('sylius_zone_member');
     }
 
     function it_builds_form_with_proper_fields(FormBuilder $builder)
     {
         $builder
-            ->add('zone', 'sylius_zone_choice', Argument::any())
-            ->shouldBeCalled()
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/spec/Form/Type/ZoneMemberTypeSpec.php]: This should stay

File: src/Sylius/Bundle/FixturesBundle/DataFixtures/ORM/LoadShippingData.php [hunk 44]
```diff
@@ -38,13 +38,13 @@ public function load(ObjectManager $manager)
         $manager->persist($heavy);
 
         $config = array('first_item_cost' => 1000, 'additional_item_cost' => 500, 'additional_item_limit' => 0);
-        $manager->persist($this->createShippingMethod(array($this->defaultLocale => 'FedEx'), 'SM1', 'USA', DefaultCalculators::FLEXIBLE_RATE, $config));
+        $manager->persist($this->createShippingMethod(array($this->defaultLocale => 'FedEx'), 'SM1', 'United States of America', DefaultCalculators::FLEXIBLE_RATE, $config));
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/FixturesBundle/DataFixtures/ORM/LoadShippingData.php]: Why not using codes here?

File: src/Sylius/Bundle/FixturesBundle/DataFixtures/ORM/LoadTaxationData.php [hunk 45]
```diff
@@ -33,13 +33,13 @@ public function load(ObjectManager $manager)
...
 
-        $taxRate = $this->createTaxRate('TR1', 'EU VAT', 'EU', 0.23);
+        $taxRate = $this->createTaxRate('TR1', 'EU VAT', 'European Union', 0.23);
         $taxRate->setCategory($taxableGoods);
 
         $manager->persist($taxRate);
         $manager->flush();
 
-        $taxableGoods->addRate($this->createTaxRate('TR2', 'US Sales Tax', 'USA', 0.08));
+        $taxableGoods->addRate($this->createTaxRate('TR2', 'US Sales Tax', 'United States of America', 0.08));
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/FixturesBundle/DataFixtures/ORM/LoadTaxationData.php]: Why not using codes here?

File: src/Sylius/Bundle/WebBundle/Resources/config/routing/backend/zone.yml [hunk 46]
```diff
@@ -11,13 +11,18 @@ sylius_backend_zone_index:
             sortable: true
 
 sylius_backend_zone_create:
-    path: /new
+    path: /type_{type}/new
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/WebBundle/Resources/config/routing/backend/zone.yml]: I think we should use `-` here `/type-{type}/new`.

File: src/Sylius/Bundle/ResourceBundle/Form/EventSubscriber/AddCodeFormSubscriber.php [hunk 47]
```diff
@@ -54,12 +54,17 @@ public function preSetData(FormEvent $event)
...
-            $disabled = (null !== $resource->getCode());
+            $disabled = null !== $resource->getCode();
         } else if (null !== $resource) {
             throw new UnexpectedTypeException($resource, CodeAwareInterface::class);
         }
 
         $form = $event->getForm();
+
+        if (!isset($resource)) {
+            $disabled = null !== $form->getConfig()->getData()['code'];
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ResourceBundle/Form/EventSubscriber/AddCodeFormSubscriber.php]: I think we don't need it anymore.

File: src/Sylius/Bundle/WebBundle/Resources/config/routing/backend/zone.yml (hunk 46 above)
Comment (NeverResponse) [inline src/Sylius/Bundle/WebBundle/Resources/config/routing/backend/zone.yml]: Actually @pjedrzejewski suggested leaving only `/{type}/new`.

File: src/Sylius/Bundle/AddressingBundle/DependencyInjection/Compiler/RegisterZoneFactoryPass.php (hunk 33 above)
Comment (NeverResponse) [inline src/Sylius/Bundle/AddressingBundle/DependencyInjection/Compiler/RegisterZoneFactoryPass.php]: :+1:

File: src/Sylius/Bundle/WebBundle/Resources/config/routing/backend/zone.yml (hunk 46 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/WebBundle/Resources/config/routing/backend/zone.yml]: :+1:

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneType.php (hunk 38 above)
Comment (NeverResponse) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneType.php]: This gonna give me headaches, but alright.

File: src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceCodeChoiceType.php (hunk 36 above)
Comment (NeverResponse) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/ProvinceCodeChoiceType.php]: Unnecessary in this case as it extends the former, so it's the same resolver.

**Comment by pjedrzejewski** [comment]
Thank you Jasiek, great work!
